### ISSUE TITLE
Error Handling

### ISSUE THREAD

**Issue by amastaneh** [TelegramBots/telegram.bot#77]
We have some errors result on `SendTextMessage` which make throw error exceptions and make slow processes. Is there any solution we safe receive these errors.
https://core.telegram.org/method/messages.sendMessage#return-errors

```
Exception Message: [Error]: PEER_ID_INVALID
```

We have [403 FORBIDDEN](https://core.telegram.org/api/errors#403-forbidden) exception when we `SendTextMessage` to user which blocked the bot.
https://core.telegram.org/api/errors

```
Exception Message: Response status code does not indicate success: 403 (Forbidden).
```

**Comment by MrRoundRobin** [comment]
Hi,

the links are specific to the normal telegram api, not the bot api.
I'll need to test the case when a bot is blocked, stay tuned...

**Comment by amastaneh** [comment]
Scenario 1
We `SendTextMessage` from a bot to a group, right after removing from group.

```json
{
  "ok": false,
  "error_code": 403,
  "description": "[Error]: Forbidden: bot was kicked from the group chat"
}
```

![image](https://cloud.githubusercontent.com/assets/6195199/12612280/d2252558-c506-11e5-8403-be1fd8c06199.png)

**Comment by amastaneh** [comment]
Scenario 2
We `SendTextMessage` from a bot to a unknown channel

```json
{
  "ok": false,
  "error_code": 400,
  "description": "[Error]: Bad Request: channel not found"
}
```

![image](https://cloud.githubusercontent.com/assets/6195199/12612351/4a7b9910-c507-11e5-80cc-7a2c9915968c.png)

**Comment by amastaneh** [comment]
Scenario 3
We `SendTextMessage` from a bot to a user, right after stop and blocking bot by this user

```json
{
  "ok": false,
  "error_code": 403,
  "description": "[Error]: Bot was blocked by the user"
}
```

![image](https://cloud.githubusercontent.com/assets/6195199/12612539/c856a59a-c508-11e5-9429-94dce33708fb.png)

**Comment by MrRoundRobin** [comment]
These errors should be thrown as an ApiException.

Can you catch these?

**Comment by amastaneh** [comment]
> Exception Handling is a heavy and expensive operation as far as performance is concerned  [[ref](http://stackoverflow.com/questions/8687113)]

We need receive these errors normally (not with an exception) and make decision based on these received messages. Anyway you can find out more respectful documents about error handling with **Condition Checking** is so better than **Exception Handling**

I think `SendTextMessage` need an output from `ResultMessage` class like 

````C#
public class ResultMessage
{
    bool IsOk {get; set;}
    Message? Result {get; set;}
    Error? Error {get; set;}
}

public class Error
{
    int Code {get; set;}
    string Description {get; set;}
}
````

**Comment by MrRoundRobin** [comment]
ok, i need to think about this

**Comment by Bouki6** [comment]
> These errors should be thrown as an ApiException.
> 
> Can you catch these?

i also test it. i can catch any exemptions with ApiException. if there was a exception we catch it like this:
```
[(Message: System.AggregateException: One or more errors occurred. ---> System.Net.Http.HttpRequestException: Response status code does not indicate success: 403 (Forbidden).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Telegram.Bot.Api.<SendWebRequest>d__78`1.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at TelegramGateWay.ViaTelegram.SendText() in C:\Users\Admin\Desktop\Projects\SimSend_SVN\trunk\TelegramGateWay\TelegramGateWay\Intermediate\ViaTelegram.cs:line 58
---> (Inner Exception #0) System.Net.Http.HttpRequestException: Response status code does not indicate success: 403 (Forbidden).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Telegram.Bot.Api.<SendWebRequest>d__78`1.MoveNext()<---
)]
funny thing is in previous releases you return the error description but in new version you just return the error code. i hope you can fix it soon. thank you.
```

**Comment by onsamp** [comment]
bot.sendMessage('@medve_dev', 'test BOt');
send to username response 

Unhandled rejection Error: 400 {"ok":false,"error_code":400,"description":"[Error]: Bad Request: channel not found"}

**Comment by mhverbakel** [comment]
I think that the implementation without throwing exceptions would result in a silent fail for most users. As this is undesirable, I am against changing to condition checking instead of exception handling.

On the other hand, it would be a good idea to include the error description in the `ApiException` class.

**Comment by a-a-k** [comment]
When async methods like `SendTextMessageAsync()` are used standalone (without user request) then bot will be frozen if he was stopped or blocked by user. Sync methods like `SendTextMessage()` works correct in this case.
How can I resolve this?

**Comment by mhverbakel** [comment]
Dear a-a-k,

First of all, for a new issue, open a new issue.

Secondly, the two functions you are referring to are literally the same. The non `Async` methods call the `Async` methods without doing anything. This is probably an issue in your own code.

If you make a new issue and share your code there, we might be able to help you.


Kind regards,
Martijn Verbakel

**Comment by a-a-k** [comment]
Dear Matijn, thank you for your response, but what does it mean?

> The non `Async` methods call the `Async` methods without doing anything. 

And I think that this is not new issue conversely this is exactly not handled error problem. Do you not think so? I'm really don't understand what you meant.

**Comment by Olfi01** [comment]
@a-a-k this might be a bit late, but the normal method just looks like this:
`public void SendTextMessage(params)
{
SendTextMessageAsync(params);
}

so its not a difference at all.
and the difference is, if you don't wait for the async method to finish with .Wait(), the exception won't be thrown, so i assume you probably DO wait for them and thus get the errors they fire.

**Comment by Olfi01** [comment]
And @amastaneh : i don't think changing the errors to statuses that have to be checked first is an option. After all, this is just a c# implementation for the API and not like a whole control panel for a bot that takes over everything for you.

**Comment by amastaneh** [comment]
Dear @Olfi01 & @mhverbakel 
Thank you for your attention and try to help. The issue is about Condition Checking versus Exception Handling

Google: [**Condition Checking** vs **Exception Handling**](https://www.google.com/search?q=Condition+Checking+vs+Exception+Handling)

1- I think logic and performance come with condition checking. 
2- Also, In sending mass messages in a loop (or in other similar scenarios) exception after a simple `SendMessage` make stop the routine process. But return a simple result when the bot is blocked from some users (or groups) is better than throw. Some of programs need continue to sending without exception interrupts.
3- The blocking result or any other situations are simple conditions. **Did you think we need throw exception when we face with some conditions or some situations** at all?

**Comment by Olfi01** [comment]
@amastaneh i think we should indeed throw exeptions. if you really want conditions, just catch those by yourself. Making it conditions would, as @mhverbakel said, result in a silent fail for those user who *arent* capable of bulding themselves a workaround. Don't you think it would be better to leave them able to determine the cause of whats happening than you not having to code a single try-catch-loop? if youre sending mass messages, just do it like for example
`while(condition)
{
try
{
//send Message
}
catch
{
//handle Exception
}
}`
Since I myself used this library even back since I was a little noob, I prefer simplicity before logical correctness.
